contracts/bitizens/AvatarItemToken.sol

```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
 
 contract AvatarItemToken is ERC721ExtendToken, AvatarItemService, AvatarChildService {
 
-  event TokenBurned(address owner, uint256 tokenId, uint256 _totalBurned);
-
-  struct AvatarItem {
-    // the listed number, the number is from the content master list
-    uint32 listNumber;
-    // set number, max is 99      
-    uint16 setNumber;
-    // true  => this itme is for avatar
-    // false => this item not for avatar
-    bool isBitizenItem;
-    // flase => This item is not part of a set    
-    // true => This item is a part of a set
-    bool isSet; 
-    // 00 => clothes  
-    // 01 => pet 
-    // type may be add in the future
-    uint8 types;
-    //01 => Common
-    //02 => Uncommon  
-    //03 => Rare      
-    //04 => Epic 
-    //05 => Legendary 
-    //06 => Godlike   
-    //07 => Unique
-    uint8 rarity;  
-    // 01 => Head
-    // 02 => Top
-    // 03 => Bottom
-    // 04 => Feet           
-    // 05 => Trinket      
-    // 06 => Acc  
-    // ...=> more 
-    uint8 socket;
-    // 00 => Male     
-    // 01 => Female     
-    // 10 => Male-only      
-    // 11 => Female-only
-    uint8 gender;
-    // extProps for future 
-    uint8 ext1;
-    // extProps for future
-    uint8 ext2;
-  }
-
-  AvatarItem[] avatarItems;
-
-  mapping(uint256 => AvatarItem) burnTokens;
-
-  uint256[] burnTokenIds;
-
-  mapping(uint256 => bool) isBurnByTokenId;
+  struct Item {
+    address foundedBy;    // item founder
+    address createdBy;    // item creator
+    bool isBitizenItem;   // true for avatar false for other
+    int16 probability;    // < 0 , decrease the mine time, > 0 increase get rare item, range to -10000 ~ 10000 mean -100.00% ~ 100.00%
+    uint256 node;         // node token id 
+    uint256 listNumber;   // list number
+    uint256 setNumber;    // set number
+    uint256 quality;      // quality of item 
+    uint8 rarity;         // 01 => Common 02 => Uncommon  03 => Rare  04 => Epic 05 => Legendary 06 => Godlike 10 => Limited
+    uint8 socket;         // 01 => Head   02 => Top  03 => Bottom  04 => Feet  05 => Trinket  06 => Acc  07 => Props
+    uint8 gender;         // 00 => Male   01 => Female 10 => Male-only 11 => Female-only  Unisex => 99
+    uint8 energy;         // increases extra mining times
+    uint8 ext;            // extra attribute for future
+  }
+
+  // item id index
+  uint256 internal tokenIndex = 0;
+  // tokenId => item
+  mapping(uint256 => Item) internal tokenInfo;
+  // all the burned token ids
+  uint256[] internal burnedTokenIds;
+  // check token id => isBurned
+  mapping(uint256 => bool) internal isBurnedToken;
+  //hash(item) => tokenIds
+  mapping(bytes32 => uint256[]) internal sameItemTokenIds;
```

Comment:
Would be great to have a brief description of what this mapping is for in the comment.
Why do you need to store an array of the same items?

Reply:
For items deep have a summary some special item count

Reply:
If you only need a sum, would be sufficient to store a `mapping(byte32 => uint) sameTokenCount`?

Reply:
I made an extension so that the same tokens could be found more easy.